CONTRIBUTING.md

```diff
@@ -1,10 +1,34 @@
 # Contributing
 
-We have issues labeled as
-[Good First Issue](https://github.com/astral-sh/uv/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22)
-and
-[Help Wanted](https://github.com/astral-sh/uv/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)
-which are good opportunities for new contributors.
+## Finding ways to help
+
+We label issues that would be good for a first time contributor as
+[`good first issue`](https://github.com/astral-sh/uv/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22).
+These usually do not require significant experience with Rust or the uv code base.
+
+We label issues that we think are a good opportunity for subsequent contributions as
+[`help wanted`](https://github.com/astral-sh/uv/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22).
+These require varying levels of experience with Rust and uv. Often, we want to accomplish these
+tasks but do not have the resources to do so ourselves.
+
+You don't need our permission to start on an issue we have labeled as appropriate for community
+contribution as described above. However, it's a good idea to indicate that you are going to work on
+an issue to avoid concurrent attempts to solve the same problem.
+
+Please check in with us before starting work on an issue that has not been labeled as appropriate
+for community contribution. We're happy to receive contributions for other issues, but it's
+important to make sure we have consensus on the solution to the problem first.
+
+Outside of issues with the labels above, issues labeled as
+[`bug`](https://github.com/astral-sh/uv/issues?q=is%3Aopen+is%3Aissue+label%3A%22bug%22) are the
+best candidates for contribution. In contrast, issues labeled with `needs-decision` or
+`needs-design` are _not_ good candidates for contribution. Please do not open pull requests for
+issues with these labels.
+
+Please do not open pull requests for new features without prior discussion. While we appreciate
+exploration of new features, we will almost always close these pull requests immediately. Adding a
+new feature to uv creates a long-term maintenance burden and requires strong consensus from the uv
+team before it is appropriate to begin work on an implementation.
```

Comment:
I think for Ruff/ty I'd probably go for slightly softer language here (since even a new lint rule counts as a "feature" in Ruff, and we generally don't insta-close those) -- something like this?

```suggestion
Please do not open pull requests for large new features without prior discussion. While we appreciate
exploration of new features, we will often close these pull requests immediately. Adding a
large new feature to uv creates a long-term maintenance burden and requires strong consensus from the uv
team before it is appropriate to begin work on an implementation.
```

But I can also see that the concerns might be different in the different projects -- we might want slightly different language in the equivalent sections in different repos, and that seems fine

Reply:
Yeah I think I'd prefer to leave this as-is for uv but it makes sense to modify for ty and ruff.

Reply:
I think we're strict enough about adding new lint rules that I highly recommend opening an issue before creating a PR. We should probably be more strict about closing PRs for Ruff.